internal/model/target_graph.go

```diff
@@ -0,0 +1,119 @@
+package model
+
+import "fmt"
+
+type TargetGraph struct {
+	// Targets in topological order.
+	sortedTargets []TargetSpec
+	byID          map[TargetID]TargetSpec
+}
+
+func NewTargetGraph(targets []TargetSpec) (TargetGraph, error) {
+	sortedTargets, err := TopologicalSort(targets)
+	if err != nil {
+		return TargetGraph{}, err
+	}
+
+	return TargetGraph{
+		sortedTargets: sortedTargets,
+		byID:          MakeTargetMap(sortedTargets),
+	}, nil
+}
+
+// In Tilt, Manifests should always be DAGs with a single root node
+// (the deploy target). This is just a quick sanity check to make sure
+// that's true, because many of our graph-traversal algorithms won't work if
+// it's not true.
+func (g TargetGraph) IsSingleSourceDAG() bool {
+	seenIDs := make(map[TargetID]bool, len(g.sortedTargets))
+	lastIdx := len(g.sortedTargets) - 1
+	for i := lastIdx; i >= 0; i-- {
+		t := g.sortedTargets[i]
+		id := t.ID()
+		isLastTarget := i == lastIdx
+		if !isLastTarget && !seenIDs[id] {
+			return false
+		}
+
+		for _, depID := range t.DependencyIDs() {
+			seenIDs[depID] = true
+		}
+	}
+	return true
+}
+
+// Visit t and its dependencies in depth-first order.
+func (g TargetGraph) VisitTree(root TargetSpec, visit func(dep TargetSpec) error) error {
+	visitedIDs := make(map[TargetID]bool)
+
+	// pre-declare the variable, so that this function can recurse
+	var helper func(current TargetSpec) error
+	helper = func(current TargetSpec) error {
+		deps, err := g.DepsOf(current)
+		if err != nil {
+			return err
+		}
+
+		for _, dep := range deps {
+			if visitedIDs[dep.ID()] {
+				continue
+			}
+
+			err := helper(dep)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = visit(current)
+		if err != nil {
+			return err
+		}
+		visitedIDs[current.ID()] = true
+		return nil
+	}
+
+	return helper(root)
+}
+
+func (g TargetGraph) DepsOf(t TargetSpec) ([]TargetSpec, error) {
```

Comment:
so this, like `t.DependencyIDs()`, is only getting the DIRECT deps of this target? and the only difference btwn this and `t.DependencyIDs()` is that this also verifies that the dep is in the graph?

Reply:
t.DependencyIDs() can't give you the TargetSpec